model/renderNativeWindow.cc

```diff
@@ -54,41 +63,69 @@ using namespace ecolab;
 namespace minsky
 {
 #ifdef USE_WIN32_SURFACE
-  inline cairo::SurfacePtr nativeWindowSurface(unsigned long window)
-  {/* TODO */}
+  inline cairo::SurfacePtr createNativeWindowSurface(WindowInformation &winInfo)
+  { /* TODO */
+  }
 #elif defined(MAC_OSX_TK)
-  inline cairo::SurfacePtr nativeWindowSurface(unsigned long window)
-  {/* TODO */}
+  inline cairo::SurfacePtr createNativeWindowSurface(WindowInformation &winInfo)
+  { /* TODO */
+  }
 #else
+  inline cairo::SurfacePtr createNativeWindowSurface(WindowInformation &wi)
+  {
+    cairo::SurfacePtr childSurface(new cairo::Surface(cairo_xlib_surface_create(wi.getDisplay(), wi.getChildWindowId(), wi.wAttr.visual, wi.childWidth, wi.childHeight), wi.childWidth, wi.childHeight));
+    cairo_surface_set_device_offset(childSurface->surface(), -wi.wAttr.x, -wi.wAttr.y);
+    return childSurface;
+  }
+#endif
 
-  int throwOnXError(Display*, XErrorEvent* ev)
+  RenderNativeWindow::~RenderNativeWindow()
   {
-    char errorMessage[256];
-    XGetErrorText(ev->display, ev->error_code, errorMessage, sizeof(errorMessage));
-    throw runtime_error(errorMessage);
+    if (this->winInfo)
+    {
+      delete this->winInfo;
+    }
```

Comment:
This still won't do. RenderNativeWindow needs to be DCAS, for Classdesc reflection reasons.